### PR TITLE
Add soundfile compatibility backend 

### PR DESCRIPTION
**Pull request by mthrok**
As a part of the "sox" backend sunset plan (#903), we add a "soundfile" backend that is compatible with the "sox_io" backend.

This PR contains;
 - The implementation of the new "soundfile" backend.
 - The function and global flag to switch the behavior of "soundfile" backend.
    No new public backend is added. We provide a switch to change the interface/behavior of "soundfile" backend.
 - Test for the new backend.

The default behavior of "soundfile" backend is not changed. The users who want to opt-in the new "soundfile" interface can do so by `torchaudio.backend.utils.use_soundfile_legacy_interface(False)` before changing the backend to "soundfile".

In 0.8.0 release, the "soundfile" backend will use this interface by default, and users can still use the legacy one with `torchaudio.backend.utils.use_soundfile_legacy_interface(True)`. In 0.9.0, the legacy interface is removed and `use_soundfile_legacy_interface` function becomes no-op, (and eventually removed from the code base.)

TODO: 
- [ ] Update #903

### CODE REVIEW

File: torchaudio/backend/utils.py [hunk 1]
```diff
@@ -33,6 +35,16 @@ def list_audio_backends() -> List[str]:
     return backends
 
 
+_USE_SOUNDFILE_LEGACY_INTERFACE = True
```
Comment (cpuhrsch) [inline torchaudio/backend/utils.py:38]: Alternatively (and probably saver) you can do
```
torchaudio._USE_SOUNDFILE_LEGACY_INTERFACE=True
```
and create a _USE_SOUNDFILE_LEGACY_INTERFACE in the ```torchaudio/__init__.py``` file.

Comment (cpuhrsch) [reply torchaudio/backend/utils.py:38]: What about someone doing

```
def fn1():
    import torchaudio
    torchaudio.backend.utils.use_soundfile_legacy_interface(True)
    def _fn():
         torchaudio.set_backend('soundfile')
         torchaudio.<do_something>

def fn2(
    import torchaudio
    torchaudio.backend.utils.use_soundfile_legacy_interface(False)

fn = fn1()
fn2()
fn() # Is this using the legacy or non-legacy version?
```

Maybe this is fixed if instead of using the global, you create a variable in ```torchaudio/__init__.py```?

File: torchaudio/backend/utils.py [hunk 2]
```diff
@@ -33,6 +35,16 @@ def list_audio_backends() -> List[str]:
...
+
+
+def use_soundfile_legacy_interface(value: bool):
+    """Switch soundfile backend interface.
+    """
+    global _USE_SOUNDFILE_LEGACY_INTERFACE
```
Comment (cpuhrsch) [inline torchaudio/backend/utils.py:44]: I'm not even sure you need to use the ```global``` keyword here

Comment (cpuhrsch) [reply torchaudio/backend/utils.py:44]: You might also want a getter and a setter to help send warnings when this variable is being changed a lot

**Review by cpuhrsch — APPROVED** [review]

File: torchaudio/backend/utils.py (hunk 1 above)
Comment (mthrok) [reply torchaudio/backend/utils.py:38]: Okay, so it seems that accessing the variable directly has more advantage (like users can check the current configuration), so I will move to it.

Comment (mthrok) [reply torchaudio/backend/utils.py]: Moved "USE_SOUNDFILE_LEGACY_INTERFACE" to `torchaudio.USE_SOUNDFILE_LEGACY_INTERFACE` and got rid of the helper function.

**Comment by vincentqb** [comment]
> In 0.8.0 release, the "soundfile" backend will use this interface by default, and users can still use the legacy one with `torchaudio.backend.utils.use_soundfile_legacy_interface(True)`. In 0.9.0, the legacy interface is removed and `use_soundfile_legacy_interface` function becomes no-op, (and eventually removed from the code base.)

I'll edit the description to reflect the correct interface:
```python
torchaudio.backend.utils.use_soundfile_legacy_interface = False
```

**Comment by mthrok** [comment]
> > In 0.8.0 release, the "soundfile" backend will use this interface by default, and users can still use the legacy one with `torchaudio.backend.utils.use_soundfile_legacy_interface(True)`. In 0.9.0, the legacy interface is removed and `use_soundfile_legacy_interface` function becomes no-op, (and eventually removed from the code base.)
> 
> I'll edit the description to reflect the correct interface:
> 
> ```python
> torchaudio.backend.utils.use_soundfile_legacy_interface = False
> ```

It's `torchaudio.USE_SOUNDFILE_LEGACY_INTERFACE`.
